pkg/roachprod/failureinjection/failures/disk_stall.go

```diff
@@ -0,0 +1,466 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package failures
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/roachprod"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+	"github.com/cockroachdb/errors"
+)
+
+var cockroachIOController = filepath.Join("/sys/fs/cgroup/system.slice", install.VirtualClusterLabel(install.SystemInterfaceName, 0)+".service", "io.max")
+
+// CGroupDiskStaller is a failure mode that stalls disk I/O using cgroups.
+// N.B. due to quirks with cgroups v2, if more than one Inject is called
+// concurrently, the first Restore call will restore all Inject failures.
+type CGroupDiskStaller struct {
+	GenericFailure
+}
+
+func MakeCgroupDiskStaller(clusterName string, l *logger.Logger, secure bool) (FailureMode, error) {
+	c, err := roachprod.GetClusterFromCache(l, clusterName, install.SecureOption(secure))
+	if err != nil {
+		return nil, err
+	}
+	genericFailure := GenericFailure{c: c, runTitle: CgroupsDiskStallName}
+	return &CGroupDiskStaller{GenericFailure: genericFailure}, nil
+}
+
+const CgroupsDiskStallName = "cgroup-disk-stall"
+
+func registerCgroupDiskStall(r *FailureRegistry) {
+	r.add(CgroupsDiskStallName, DiskStallArgs{}, MakeCgroupDiskStaller)
+}
+
+func (s *CGroupDiskStaller) Description() string {
+	return CgroupsDiskStallName
+}
+
+type DiskStallArgs struct {
+	StallLogs   bool
+	StallReads  bool
+	StallWrites bool
+	// If true, allow the failure mode to restart nodes as needed. E.g. dmsetup requires
+	// the cockroach process to not be running to properly setup. If RestartNodes is true,
+	// then the failure mode will restart the cluster for the user.
+	RestartNodes bool
```

Comment:
The current disk stall roachtests are okay with the node dying since they don't try to do anything else in the test after the fact. However, I still think we should restart the node at the end so artifacts can be collected - you can see in all of the current roachtests there's no cockroach logs for the stalled node.